master/internal/api_experiment.go

```diff
@@ -0,0 +1,168 @@
+	}
+	return nil
+}
+
+func canCancel(dbExp *model.Experiment) error {
+	if _, exist := model.RunningStates[dbExp.State]; !exist {
+		return status.Error(codes.InvalidArgument, "experiment is not in a running state")
+	}
+	return nil
+}
+
+func canArchive(dbExp *model.Experiment) error {
+	if dbExp.Archived {
+		return status.Error(codes.InvalidArgument, "experiment is already archived")
+	}
+	return nil
+}
+
+func canUnArchive(dbExp *model.Experiment) error {
+	if !dbExp.Archived {
+		return status.Error(codes.InvalidArgument, "experiment is not archived")
+	}
+	return nil
+}
+
+type validityCheck func(a *model.Experiment) error
+
+func isValidOperation(dbConn *db.PgDB, expID int, checker validityCheck) (*model.Experiment, error) {
+
+	dbExp, err := dbConn.ExperimentByID(expID)
+
+	switch {
+	case err == db.ErrNotFound:
+		return nil, errUserNotFound
+	case err != nil:
+		return nil, errors.Wrapf(err, "loading experiment %v", expID)
+	}
+
+	err = checker(dbExp)
+	if err != nil {
+		return nil, err
+	}
+
+	return dbExp, nil
+}
+
+func (a *apiServer) ActivateExperiment(
+	ctx context.Context, req *apiv1.SetExperimentAttributeRequest) (*apiv1.SetExperimentAttributeResponse, error) {
+
+	expID := (int)(req.Id) // platform dependent int
+
+	_, err := isValidOperation(a.m.db, expID, canActivate)
+	if err != nil {
+		return nil, err
+	}
+
+	a.m.system.TellAt(actor.Addr("experiments", expID), model.ActiveState)
+
+	resultingExperiment := experimentv1.Experiment{Id: req.Id}
+	return &apiv1.SetExperimentAttributeResponse{Experiment: &resultingExperiment}, nil
+}
+
+func (a *apiServer) PauseExperiment(
+	ctx context.Context, req *apiv1.SetExperimentAttributeRequest) (*apiv1.SetExperimentAttributeResponse, error) {
+	expID := (int)(req.Id)
+
+	_, err := isValidOperation(a.m.db, expID, canPause)
+	if err != nil {
+		return nil, err
+	}
+
+	a.m.system.TellAt(actor.Addr("experiments", expID), model.PausedState)
+
+	resultingExperiment := experimentv1.Experiment{Id: req.Id}
+	return &apiv1.SetExperimentAttributeResponse{Experiment: &resultingExperiment}, nil
+}
+
+func (a *apiServer) CancelExperiment(
+	ctx context.Context, req *apiv1.SetExperimentAttributeRequest) (*apiv1.SetExperimentAttributeResponse, error) {
+	expID := (int)(req.Id)
+
+	_, err := isValidOperation(a.m.db, expID, canCancel)
+	if err != nil {
+		return nil, err
+	}
+
+	a.m.system.TellAt(actor.Addr("experiments", expID), model.StoppingCanceledState)
+
+	resultingExperiment := experimentv1.Experiment{Id: req.Id}
+	return &apiv1.SetExperimentAttributeResponse{Experiment: &resultingExperiment}, nil
+}
+
+func (a *apiServer) KillExperiment(
+	ctx context.Context, req *apiv1.SetExperimentAttributeRequest) (*apiv1.SetExperimentAttributeResponse, error) {
+	expID := (int)(req.Id)
+
+	_, err := isValidOperation(a.m.db, expID, canCancel)
+	if err != nil {
+		return nil, err
+	}
+
+	a.m.system.TellAt(actor.Addr("experiments", expID), model.StoppingCanceledState)
```

Comment:
blocking: We should change this to an ask request to ensure that the object exists in memory. We should also do the appropriate error handling if it is not.